lib/Mojo/IOLoop/Delay.pm

```diff
@@ -3,37 +3,130 @@ use Mojo::Base 'Mojo::EventEmitter';
 
 use Mojo::IOLoop;
 use Mojo::Util;
+use Scalar::Util qw(blessed weaken);
 
 has ioloop    => sub { Mojo::IOLoop->singleton };
 has remaining => sub { [] };
 
+sub all {
+  my @promises = @_;
+
+  my $all = $promises[0]->_clone;
+
+  my $results   = [];
+  my $remaining = scalar @promises;
+  for my $i (0 .. $#promises) {
+    $promises[$i]->then(
+      sub {
+        $results->[$i] = [@_];
+        $all->resolve(@$results) if --$remaining <= 0;
+      },
+      sub { $all->reject(@_) },
+    );
+  }
+
+  return $all;
+}
+
 sub begin {
   my ($self, $offset, $len) = @_;
   $self->{pending}++;
   my $id = $self->{counter}++;
   return sub { $self->_step($id, $offset // 1, $len, @_) };
 }
 
+sub catch { shift->then(undef, shift) }
```

Comment:
This breaks documented usage and can issue false negatives in common practice. I feel that to change this without a major version number increment or a deprecation is dangerous to our users. This could be mitigated by wrapping the passed in callback. Whether that be in a deprecation cycle or just a long-term fix is immaterial to me.